### PR TITLE
Fix: incorrect avatar raycasting

### LINKED ISSUE #3859: [QA] Functionality issues with the Xmas area restricting passport usage
**Description:**
Due to increased range at which the passports can be opened (most likely a change associated with the implementation of avatar outliner), the users are able to open passports even in area which restricts them.
In order to reproduce the issue, the user needs to aim at an avatar positioned at a distance in the Xmas area's restricted zone and then open the passport. Moreover, reproducing the issue enables passport opening for the entire area. The targeted avatar does not have to be outside of the restricted zone in order to reproduce the issue.

**Steps to reproduce:**

1. Load into Decentraland on the RC branch (28.12.2022) as a guest or via a wallet.
2. Reach the Xmas area which blocks the use of passports (-116, 106).
3. Aim at a user located at a distance.
4. Observe the issue.

**Reproduction rate:**
5/5

**Results:**
Passports can be opened inside the zone which disables their usage.

**Expected results:**
Passports cannot be opened in the area restricting passport usage.

**Notes:**

- The issue can be reproduced on both the guest and wallet accounts.

- The issue seems to occur only for the Xmas area.

- The issue was reproduced on the following release branch: [RC Branch](https://play.decentraland.org/?renderer-branch=release/release-20221228).

- Please see the attached video for reference: https://images.zenhubusercontent.com/337227404/abcd4a53-ff17-4698-881f-df34d67f0cbf/_3859.mp4.

- Platform: Web client, Windows client.

### PR DESCRIPTION
**Pull request by mikhail-dcl**
## What does this PR change?

Closes #3859

**Definition of problem:**
Raycasting for passport is limited by range, while raycasting for outlining is unlimited. But they both were handled by a single piece of code that was very coupled. 

**Approach**
Detach avatar outlining from other functionality.

**Solution**
`PointerEventsController` is junk and without refactoring, it was impossible to approach properly. I had to extract all the functionality related to hovering to a separate class:
- The logic there was unclear so I modified it to be more consistent
- I tried to preserve the previous behavior
- Add an independent control of hovering so now there can be different `IPointerEvent`s with completely independent logic on the same object, and it will work correctly.
- It allowed me to introduce `AvatarOutlineOnHoverEvent` that handles outline detection only and move there a piece of logic from `AvatarShape` (that was in the wrong domain anyway)

## How to test the changes?

- Test Avatar and other objects Input (clicks, up, down, hovering): I had to modify the system behind them so I may expect some hiccups. 
- Behavior of Passport raycasting and Avatar Outling should work as expected regardless of the feature flag.

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarOnPointerDown/AvatarOutlineOnHoverEvent.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+using DCLPlugins.UUIDEventComponentsPlugin.UUIDComponent.Interfaces;
+using UnityEngine;
+
+namespace DCL.Components
+{
+    /// <summary>
+    /// Complementary class to detect avatar outlining only,
+    /// It does not live on its own and should coexist with `AvatarOnPointerDown`
+    /// </summary>
+    public class AvatarOutlineOnHoverEvent : MonoBehaviour, IUnlockedCursorInputEvent, IPoolLifecycleHandler
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarOnPointerDown/AvatarOutlineOnHoverEvent.cs:12]: i dont really feel like naming this class as an Event, perhaps a handler

**Review by lorux0 — APPROVED** [review]

File: unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/PointerEventsController/PointerEventsController.cs [hunk 2]
```diff
@@ -17,52 +17,45 @@ namespace DCL
...
+        private readonly InputController_Legacy inputControllerLegacy;
 
         private DataStore_ECS7 dataStoreEcs7 = DataStore.i.ecs7;
 
         public PointerEventsController(InputController_Legacy inputControllerLegacy,
             InteractionHoverCanvasController hoverCanvas)
         {
             this.inputControllerLegacy = inputControllerLegacy;
-            this.hoverCanvas = hoverCanvas;
+            pointerHoverController = new PointerHoverController(inputControllerLegacy, hoverCanvas);
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/PointerEventsController/PointerEventsController.cs:44]: I think it's not needed here as it's not gonna be exposed publicly

File: unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/PointerEventsController/PointerHoverController.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            pointerInputEvent.entity == colliderInfo.entity;
+
+        public void ResetHoveredObject()
+        {
+            if (currentHoveredObjectInfo.GameObject != null)
+            {
+                foreach (var listener in currentHoveredObjectInfo.ActiveListeners)
+                {
+                    // Respect Unity GO lifecycle
+                    if (Equals(listener, null))
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/PointerEventsController/PointerHoverController.cs:168]: It is not possible as it's an interface

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarOnPointerDown/AvatarOutlineOnHoverEvent.cs (hunk 1 above)
Comment (mikhail-dcl) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarOnPointerDown/AvatarOutlineOnHoverEvent.cs:12]: I just maintain the current way of naming. More changes should be done in a separate refactoring

File: unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/PointerEventsController/PointerHoverController.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                    OnPointerHoverStarts?.Invoke();
+                }
+            }
+        }
+
+        private void ResolveActiveListeners(float distance, ColliderInfo colliderInfo, Type typeToUse)
+        {
+            var inactiveListenersCount = currentHoveredObjectInfo.InactiveListeners.Count;
+
+            var activeListeners = ListPool<IPointerEvent>.Get();
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/PointerEventsController/PointerHoverController.cs:85]: to be able to remove from the original collection while iterating

File: unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/PointerEventsController/PointerEventsController.cs [hunk 5]
```diff
@@ -146,102 +140,20 @@ public void Update()
                 return;
             }
 
-            if (CollidersManager.i.GetColliderInfo(hitInfo.collider, out ColliderInfo info))
-                newHoveredInputEvent = (IPointerEvent)info.entity.gameObject.GetComponentInChildren(typeToUse);
-            else
-                newHoveredInputEvent = (IPointerEvent)hitInfo.collider.GetComponentInChildren(typeToUse);
+            var targetGO = CollidersManager.i.GetColliderInfo(hitInfo.collider, out var colliderInfo)
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/PointerEventsController/PointerEventsController.cs:143]: I will do it

**Review by kwssbocian — APPROVED** [review]
Issue appears to be fixed.
✔️Passports no longer appear upon interaction with user avatars in places where passports are disabled.
Additionally no other issues have been observed related to passport behavior, inputs and avatar outline.

**Pull request merged**